internal/server/data/query.go

```diff
@@ -0,0 +1,120 @@
+package data
+
+import (
+	"strings"
+
+	"github.com/infrahq/infra/uid"
+)
+
+type queryBuilder struct {
+	query strings.Builder
+	Args  []interface{}
+}
+
+// Query initializes a queryBuilder and returns it populated with stmt.
+// The stmt string can generally contain any SELECT, FROM, or JOIN
+// clauses, and may contain the entire query as long as there are no
+// query parameters.
+// Use queryBuilder.B to add sections of a query with parameters.
+func Query(stmt string) *queryBuilder {
+	q := &queryBuilder{}
+	q.query.WriteString(stmt + " ")
+	return q
+}
+
+// B adds clause and args to the query. The clause must be a trusted string
+// literal. Any arguments must be passed as args so that they are properly
+// escaped by the database driver.
+func (q *queryBuilder) B(clause string, args ...interface{}) {
+	q.query.WriteString(clause + " ")
+	q.Args = append(q.Args, args...)
+}
+
+// String returns the query string, which is used as the first parameter to
+// WriteTxn.Exec, or ReadTxn.Query. You must also pass q.Args as the varargs.
+func (q *queryBuilder) String() string {
+	return q.query.String()
+}
+
+type Table interface {
+	Table() string
+	// Columns returns the names of the tables columns.
+	Columns() []string
+}
+
+type Insertable interface {
+	Table
+	// Values returns the values for all fields. The values must be in the same
+	// order as the column names returned by Columns.
+	Values() []any
+	// OnInsert is called by insert to initialize values before inserting.
+	OnInsert() error
+}
+
+type Updatable interface {
+	Insertable
+	// Primary returns the value for the field that is mapped to the primary key
+	// of the table.
+	Primary() uid.ID
+	// OnUpdate is called by update to initialize values before updating.
+	OnUpdate() error
+}
+
+type Deletable interface {
+	Table() string
+	Primary() uid.ID
+}
+
+type Selectable interface {
+	Table
+	// ScanFields returns pointers to all the fields, which should be used in
+	// sql.Rows.Scan. The fields must be in the same order as the column names
+	// returned by Columns.
+	ScanFields() []any
+}
+
+func insert(tx WriteTxn, item Insertable) error {
+	if err := item.OnInsert(); err != nil {
```

Comment:
I'd like to see the org selector in one of these functions.

Reply:
Not sure what you mean. These is an insert, so there's no selector. I do call `setOrg` just below here, which has the panic I think you are looking for.

`update` has to operate on a primary ID, so there's nothing to really validate. Any already existing DB item must already have an OrgID.